src/JustSaying/Messaging/Channels/ChannelConsumer.cs

```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.AwsTools.MessageHandling.Dispatch;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal class ChannelConsumer : IChannelConsumer
+    {
+        private IAsyncEnumerable<IQueueMessageContext> _messageSource;
+        private readonly IMessageDispatcher _dispatcher;
+
+        public ChannelConsumer(IMessageDispatcher dispatcher)
+        {
+            _dispatcher = dispatcher;
+        }
+
+        public IChannelConsumer ConsumeFrom(IAsyncEnumerable<IQueueMessageContext> messageSource)
+        {
+            _messageSource = messageSource;
+            return this;
+        }
+
+        public async Task Run(CancellationToken stoppingToken)
+        {
+            await foreach (var messageContext in _messageSource.WithCancellation(stoppingToken))
+            {
+                stoppingToken.ThrowIfCancellationRequested();
```

Comment:
Open question with no right answer, just for discussion:

Should we throw the operation cancelled exception here, or just break the loop and return?

Reply:
I think I prefer this strategy for the consistency it gives us in cancellation (always throws, and we can test for it), but I'm not totally attached to how this works currently.